### ISSUE TITLE
Flow-sensitive type narrowing does not extend into lexically subsequent closures

### ISSUE THREAD

**Issue by isiahmeadows** [microsoft/TypeScript#37802]
**TypeScript Version:** 3.3.3, 3.5.1, 3.6.3, 3.7.5, 3.8.2, 3.9.0-beta, and 3.9.0-dev.20200328 (every version in the playground at the time of writing)

**Search Terms:** flow-sensitive type closure

**Expected behavior:** No type errors

**Actual behavior:** In the `return value` for `deferred`, I got this error:

```
Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'. (2322)
```

**Related Issues:** Couldn't find any

**Code**
```ts
function immediate(value?: string): string {
    if (value == null) {
        value = "";
    }

    return value;
}

function deferred(value?: string): () => string {
    if (value == null) {
        value = "";
    }

    return (): string => {
        return value;
    };
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function immediate(value) {
    if (value == null) {
        value = "";
    }
    return value;
}
function deferred(value) {
    if (value == null) {
        value = "";
    }
    return () => {
        return value;
    };
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
  }
}
```

</details>

**Playground Link:** [For nightly](https://www.typescriptlang.org/en/play?esModuleInterop=false&experimentalDecorators=false&emitDecoratorMetadata=false&jsx=0&ts=3.9.0-dev.20200328#code/GYVwdgxgLglg9mABDAtigpgExgQyugCgDccAbEdAfgC5EBnKAJxjAHMBKWh5txAbwBQiYcmCJiZCogC80xGBClS7fkJHqS5dDMQAiXQG41wgL4DjiRuighGSTRSNmBoSLASJM6YOkZXMElo09EwsHLQEKtIAfCE8rKrqMGKBUrLyisqJ6iIO2nL6RurO6lY2duKccWEysYI5ImW29pLoRSImTkA), but you can verify with other versions easily enough.

**Comment by isiahmeadows** [comment]
Just to clarify: this is not related to cases like either of these:

```ts
function deferred(value?: string): () => string {
    if (value == null) {
        value = "";
    }

    function result(): string {
        return value;
    }

    return result;
}

function deferred(value?: string): () => string {
    const result = (): string => {
        return value;
    };

    if (value == null) {
        value = "";
    }

    return result;
}
```

It's only about cases where the function is created *after* the flow-sensitive cast, like the example in the bug as well as this:

```ts
function deferred(value?: string): () => string {
    if (value == null) {
        value = "";
    }

    const result = (): string => {
        return value;
    };

    return result;
}
```

**Comment by kitsonk** [comment]
Duplicate of #11498

**Comment by RyanCavanaugh** [comment]
This is specific to bindings which are mutable, and parameters are mutable bindings. The problem is that TS doesn't do an expensive forward-pass on the code to see if there are any subsequent assignments to `value`, so it can't disavow the possibility of code like this
```ts
function deferred(value?: string): () => string {
    if (value == null) {
        value = "";
    }

    const p = (): string => {
        return value;
    };
    value = undefined;
    return p;
}
```

**Comment by RyanCavanaugh** [comment]
The hypothetical performance-minded solution here would be to allow some kind of immutability annotation on the parameter, but the example kind of defeats that. Writing `const v = value ?? ""` and using `v` thereafter would be the recommended solution
